libs/injection/bbhnet/injection/injection.py

```diff
@@ -0,0 +1,261 @@
+    snr = np.sum(snr[fmin <= data_freq])
+    snr = np.sqrt(snr)
+
+    return snr
+
+
+def generate_gw(
+    sample_params,
+    ifo,
+    waveform_generator=None,
+    get_snr=False,
+    noise_psd=None,
+    whiten_fn=None,
+):
+    """Generate gravitational-wave events
+
+    Arguments:
+    - sample_params: dictionary of GW parameters
+    - sample_duration: time duration of each sample
+    - triggers: trigger time (relative to `sample_duration`) of each sample
+    - ifo: interferometer
+    - waveform_generator: bilby.gw.WaveformGenerator with appropriate params
+    - get_snr: return the SNR of each sample
+    - noise_psd: background noise PSD used to calculate SNR the sample
+    - whiten_fn: whiten each sample using the background noise PSD
+    """
+
+    n_sample = len(sample_params["geocent_time"])
+
+    if waveform_generator is None:
+        waveform_generator = bilby.gw.WaveformGenerator(
+            duration=8,
+            sampling_frequency=16384,
+            frequency_domain_source_model=lal_binary_black_hole,
+            parameter_conversion=convert_to_lal_binary_black_hole_parameters,
+            waveform_arguments={
+                "waveform_approximant": "IMRPhenomPv2",
+                "reference_frequency": 50,
+                "minimum_frequency": 20,
+            },
+        )
+
+    sample_rate = waveform_generator.sampling_frequency
+    waveform_duration = waveform_generator.duration
+    waveform_size = int(sample_rate * waveform_duration)
+
+    signals = np.zeros((n_sample, waveform_size))
+    snr = np.zeros(n_sample)
+
+    ifo = bilby.gw.detector.get_empty_interferometer(ifo)
+    for i in range(n_sample):
+        # Get parameter for one signal
+        p = dict()
+        for k, v in sample_params.items():
+            p[k] = v[i]
+        ra, dec, geocent_time, psi = (
+            p["ra"],
+            p["dec"],
+            p["geocent_time"],
+            p["psi"],
+        )
+
+        # Generate signal in IFO
+        polarizations = waveform_generator.time_domain_strain(p)
+        signal = np.zeros(waveform_size)
+        b, a = sig.butter(
+            N=8,
+            Wn=waveform_generator.waveform_arguments["minimum_frequency"],
+            btype="highpass",
+            fs=waveform_generator.sampling_frequency,
+        )
+        for mode in polarizations.keys():
+            # Get ifo response
+            response = ifo.antenna_response(ra, dec, geocent_time, psi, mode)
+            polarizations[mode] = sig.filtfilt(b, a, polarizations[mode])
+            signal += polarizations[mode] * response
+
+        # Total shift = shift to trigger time + geometric shift
+        dt = waveform_duration / 2.0
+        dt += ifo.time_delay_from_geocenter(ra, dec, geocent_time)
+        signal = np.roll(signal, int(np.round(dt * sample_rate)))
+
+        # Calculate SNR
+        if noise_psd is not None:
+            if get_snr:
+                snr[i] = get_snr(signal, noise_psd, sample_rate)
```

Comment:
I believe the get_snr function can be array-ified and run at the end. I can take a crack at this if you'd like and submit it as a PR to your branch, which you can merge before I merge this into main.

Reply:
I'd appreciate if you could give this a try